src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+    hash_instrs[hash_id].clear();
+    hash_instrs[hash_id].insert(first);
+}
+
+instruction_ref horizontal_fusion_impl::break_split(int enum_ndx, instruction_ref split_ins)
+{
+    const operation& op = split_ins->get_operator();
+    int first           = (any_cast<op::split>(op)).slice_selector.first;
+    int second          = (any_cast<op::split>(op)).slice_selector.second;
+    if((first < 0) || (second < first))
+        MIGRAPHX_THROW("unexpected selector");
+    if((enum_ndx != first) && (enum_ndx != second))
+        MIGRAPHX_THROW("unexpected slice enumerator");
+
+    if(first == second)
+        return split_ins;
+    int axis                    = (any_cast<op::split>(op)).axis;
+    std::vector<int> slice_dims = (any_cast<op::split>(op)).slice_dims;
+    instruction_ref input       = split_ins->inputs().at(0);
+    instruction_ref new_split   = p_program->insert_instruction(
+        split_ins, op::split{axis, slice_dims, {enum_ndx, enum_ndx}}, input);
+    operation& op_edit = split_ins->get_operator_edit();
+
+    if(first == enum_ndx)
+        (any_cast<op::split>(op_edit)).slice_selector.first = enum_ndx + 1;
+    else
+        (any_cast<op::split>(op_edit)).slice_selector.second = enum_ndx - 1;
+
+    std::vector<shape> shapes;
+    shapes.push_back(input->get_shape());
+    shape new_shape = (any_cast<op::split>(op_edit)).compute_shape(shapes);
+    split_ins->set_shape(new_shape);
+    return new_split;
+}
+
+bool horizontal_fusion_impl::has_unique_output(const std::vector<instruction_ref>& instrs)
+{
+    std::unordered_map<instruction_ref, bool> seen;
+    for(auto&& ins : instrs)
+    {
+        for(auto&& output : ins->outputs())
+        {
+            if(seen.find(output) != seen.end())
+                return false;
+            seen[output] = true;
+        }
+    }
+    return true;
+}
+
+std::vector<unsigned> horizontal_fusion_impl::find_cluster(unsigned id)
+{
+    std::vector<unsigned> cluster;
+    cluster.push_back(id);
+    unsigned cur = id;
+    int size     = hash_instrs[id].size();
+    // Find a sub-tree of the hash tree to be fused together.
+    // Every node in the sub-tree contain the same amount of instructions.
+    while((hash_outputs.find(cur) != hash_outputs.end()) && (hash_outputs[cur].size() == 1))
+    {
+        unsigned output = (*(hash_outputs[cur].begin()))->id;
+        // Currently instruction can not have more than one same outputs.
+        // Therefore skip outputs that are not unique.
+        if((hash_instrs.find(output) != hash_instrs.end()) &&
+           (hash_instrs[output].size() == size) && has_unique_output(get_instrs(output)))
+        {
+            cluster.push_back(output);
+            cur = output;
+        }
+        else
+            break;
+    }
+    return cluster;
+}
+
+void horizontal_fusion_impl::transform_layers(
+    std::vector<std::vector<instruction_ref>>& all_inputs,
+    std::unordered_map<instruction_ref, instruction_ref>& root,
```

Comment:
These two parameters can be declared as const.

Reply:
fixed.